lib/ansible/modules/dnf.py

```diff
@@ -509,40 +510,31 @@ def _compare_evr(self, e1, v1, r1, e2, v2, r2):
         return rc
 
     def _ensure_dnf(self):
-        if not HAS_DNF:
-            if PY2:
-                package = 'python2-dnf'
-            else:
-                package = 'python3-dnf'
-
-            if self.module.check_mode:
-                self.module.fail_json(
-                    msg="`{0}` is not installed, but it is required"
-                    "for the Ansible dnf module.".format(package),
-                    results=[],
-                )
-
-            rc, stdout, stderr = self.module.run_command(['dnf', 'install', '-y', package])
-            global dnf
-            try:
-                import dnf
-                import dnf.cli
-                import dnf.const
-                import dnf.exceptions
-                import dnf.subject
-                import dnf.util
-            except ImportError:
-                self.module.fail_json(
-                    msg="Could not import the dnf python module using {0} ({1}). "
-                        "Please install `{2}` package or ensure you have specified the "
-                        "correct ansible_python_interpreter.".format(sys.executable, sys.version.replace('\n', ''),
-                                                                     package),
-                    results=[],
-                    cmd='dnf install -y {0}'.format(package),
-                    rc=rc,
-                    stdout=stdout,
-                    stderr=stderr,
-                )
+        if HAS_DNF:
+            return
+
+        system_interpreters = ['/usr/libexec/platform-python',
+                               '/usr/bin/python3',
+                               '/usr/bin/python2',
+                               '/usr/bin/python']
+
+        if not has_respawned():
+            # probe well-known system Python locations for accessible bindings, favoring py3
+            interpreter = probe_interpreters_for_module(system_interpreters, 'dnf')
+
+            if interpreter:
+                # respawn under the interpreter where the bindings should be found
+                respawn_module(interpreter)
+                # end of the line for this module, the process will exit here once the respawned module completes
+
+        # done all we can do, something is just broken (auto-install isn't useful anymore with respawn, so it was removed)
+        self.module.fail_json(
```

Comment:
The old code tried to install `python{2|3}-dnf` like apt but we don't seem to do that anymore which would be a breaking change.

Reply:
Yeah, without respawn, that was necessary (since you might be running the module under py2, but the builtin install was on py3). With respawn, we'll find whatever's available with the system python/dnf and use it- currently anyway, `dnf` itself uses `python(x)-dnf`, so if one of the system interpreters doesn't have that package installed (and the `dnf` Python module importable), `dnf` itself won't be any use to install it. We also want to be careful not to accidentally drag in a whole new Python trying to install that (a problem the `apt` modules solve in a slightly different way, since they still do the install)